design-documents/testing-strategy.md

```diff
@@ -0,0 +1,39 @@
+## Updated Magento Commerce Testing Strategy
+
+### Terms
+
+<!-- Describe any new terms used in the document -->
+
+### Overview
+
+<!-- Describe what we are trying to solve and a few sentence summary of design approach. -->
+
+Despite running 20,000 - 30,000 tests per pull request, we have little to no confidence in the stability of the product. This means that we still need a manual regression process that lasts weeks in the best of cases. 
+
+A sizable portion of this can be attributed to our current high-level testing mentality that prioritizes system-level black box testing over other types of tests. While these types of tests are necessary and good to have, they only provide implicit coverage to the code paths of the concrete implementations that happen to be loaded in the high-level black box tests (they are also expensive/difficult to maintain). We still need explicit coverage for the concrete implementations that extension developers have hard dependencies on. This is true for the simple code paths as well as variations both positive and negative flows.
+
+### Design
+
+#### New Testing Strategy:
+
+The new strategy for testing relies on having test coverage at each layer. The idea is to have many small and fast tests. Therefore, tests MUST be written in the following order of granularity and the higher-level the test, the fewer tests there will be:
+
+1. Code MUST be tested on its own. This MAY be done through unit or integration testing as appropriate. This level of test coverage ensures that the code can properly function on its own.  
```

Comment:
I would suggest to clarify that we want to write black box unit tests. Currently we have a lot of unit tests that are white box tests. They don't test much and break when you make changes (including those that don't affect functionality) because everything is mocked including business logic.

Reply:
I agree, we are going to emphasize that unit tests are going to be fairly rare in Magento code. Essentially code tests will default to integration tests unless there is an obvious use case for a unit test.